### ISSUE TITLE
Always show Starting DevTools

### ISSUE THREAD

**Issue by xkeyC** [flutter/flutter-intellij#5234]
Thanks for the feedback! If your issue is related to the Flutter framework itself,
please open an issue at
[github.com/flutter/flutter](https://github.com/flutter/flutter/issues/new).

## Steps to Reproduce

![image](https://user-images.githubusercontent.com/39891083/106553664-c7e0f880-6554-11eb-9517-4f10708fcb88.png)

I am using flutter in China, and I must access dart pub through http proxy, but the flutter intellij plugin does not support the use of idea proxy settings.
When I start android-studio by adding the http_proxy environment variable, flutter_intellij uses the proxy correctly, but I cannot debug and run the flutter app.

###### Use Http Proxy
```
set http_proxy=http://127.0.0.1:7890 & set https_proxy=http://127.0.0.1:7890
"C:\Program Files\Android\Android Studio\bin\studio64.exe"
```
Starting DevTools will load normally, but:
```
[        ] Waiting for observatory port to be available...
[+3036 ms] Observatory URL on device: http://127.0.0.1:37729/Emoi_L7U-s8=/
[        ] executing: C:\Users\me\AppData\Local\Android\Sdk\platform-tools\adb.exe -s emulator-5554 forward tcp:0 tcp:37729
[  +40 ms] 11498
[        ] Forwarded host port 11498 to device port 37729 for Observatory
[   +4 ms] Caching compiled dill
[  +41 ms] Connecting to service protocol: http://127.0.0.1:11498/Emoi_L7U-s8=/
[ +206 ms] Exception attempting to connect to the VM Service: WebSocketException: Connection to 'http://127.0.0.1:11498/Emoi_L7U-s8=/ws#' was not upgraded to websocket
[        ] This was attempt #1. Will retry in 0:00:00.100000.
[  +12 ms] E/DartVM  (30122): vm-service: Error: Unhandled exception:
[        ] E/DartVM  (30122): WebSocketException: Invalid WebSocket upgrade request
[        ] E/DartVM  (30122): 
[        ] E/flutter (30122): [ERROR:flutter/runtime/dart_isolate.cc(1137)] Unhandled exception:
[        ] E/flutter (30122): WebSocketException: Invalid WebSocket upgrade request
[ +109 ms] Fail to connect to service protocol: http://127.0.0.1:11498/Emoi_L7U-s8=/: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:11498/Emoi_L7U-s8=/ws
Activating Dart DevTools...
[{"event":"app.progress","params":{"appId":"14233403-292e-4903-8e05-627f62656f86","id":"2","progressId":null,"message":"Activating Dart DevTools...","finished":false}}]
[{"event":"app.progress","params":{"appId":"14233403-292e-4903-8e05-627f62656f86","id":"2","progressId":null,"finished":true}}]
[+9085 ms] Error connecting to the service protocol: failed to connect to http://127.0.0.1:11498/Emoi_L7U-s8=/
```
If use http_proxy environment variable,flutter run will Will be damaged.

### I think the best solution

flutter-intellij should read the proxy settings of the idea and attach it to the runtime environment when executing any `flutter pub.`

![image](https://user-images.githubusercontent.com/39891083/106554894-50f92f00-6557-11eb-98c1-cc48c8c7496f.png)

If execute flutter run, should not use a proxy.

![image](https://user-images.githubusercontent.com/39891083/106555023-b3eac600-6557-11eb-9c10-32c283f44c4d.png)

## Version info
flutter-intellij: 53.0.1
flutter:
```
[√] Flutter (Channel master, 1.26.0-18.0.pre.90, on Microsoft Windows [Version 10.0.19042.746], locale zh-CN)
    • Flutter version 1.26.0-18.0.pre.90 at C:\Users\me\AppData\Local\Android\flutter
    • Framework revision 25134a16d1 (3 days ago), 2021-01-29 23:52:30 -0800
    • Engine revision ce14c8a841
    • Dart version 2.12.0 (build 2.12.0-276.0.dev)

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\me\AppData\Local\Android\Sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = C:\Users\me\AppData\Local\Android\Sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[√] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[√] VS Code (version 1.52.1)
    • VS Code at C:\Users\me\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension can be installed from:
       https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[!] Proxy Configuration
    • HTTP_PROXY is set
    ! NO_PROXY is not set

[√] Connected device (3 available)
    • sdk gphone x86 64 arm64 (mobile) • emulator-5554 • android-x64    • Android 11 (API 30) (emulator)
    • Chrome (web)                     • chrome        • web-javascript • Google Chrome 88.0.4324.104
    • Edge (web)                       • edge          • web-javascript • Microsoft Edge 88.0.705.53

```

**Comment by tehprofessor** [comment]
I was getting this too and I don't use a proxy.

It looks like it was caused by the issue linked in #5230 

To fix it, I needed to:

1. `flutter channel dev`
2. Download a previous version of this pacakge: `https://plugins.jetbrains.com/files/9212/107560/flutter-intellij.zip?updateId=107560&pluginId=9212&family=INTELLIJ`
3. Install the package by going to `Extensions` and select `Install from Disk`
4. Activate the old version of devtools: `flutter pub global activate devtools 0.9.6+3`
5. In intellij go to `File > Invalidate Caches and Restart`

Another symptom I encountered was the autocomplete not working anymore. It wasn't until step 5 above that the autocomplete finally started working.

Hope that helps you. I was losing my freakin' mind over this...

**Comment by xkeyC** [comment]
> I was getting this too and I don't use a proxy.
> 
> It looks like it was caused by the issue linked in #5230
> 
> To fix it, I needed to:
> 
> 1. `flutter channel dev`
> 2. Download a previous version of this pacakge: `https://plugins.jetbrains.com/files/9212/107560/flutter-intellij.zip?updateId=107560&pluginId=9212&family=INTELLIJ`
> 3. Install the package by going to `Extensions` and select `Install from Disk`
> 4. Activate the old version of devtools: `flutter pub global activate devtools 0.9.6+3`
> 5. In intellij go to `File > Invalidate Caches and Restart`
> 
> Another symptom I encountered was the autocomplete not working anymore. It wasn't until step 5 above that the autocomplete finally started working.
> 
> Hope that helps you. I was losing my freakin' mind over this...

Thanks, this may be a solution, but I think the proxy method of flutter-intellij still needs improvement.

**Comment by kenzieschmoll** [comment]
Some links in case they are helpful @helin24 

Using Flutter in China: https://flutter.dev/community/china
Brief discussion between Jonah and I regarding China on my flutter PR to serve DevTools at app start: https://github.com/flutter/flutter/pull/73366#discussion_r553012943

**Comment by xkeyC** [comment]
> Some links in case they are helpful @helin24
> 
> Using Flutter in China: https://flutter.dev/community/china
> Brief discussion between Jonah and I regarding China on my flutter PR to serve DevTools at app start: [flutter/flutter#73366 (comment)](https://github.com/flutter/flutter/pull/73366#discussion_r553012943)

Thanks,but I'm using the master channel,the China mirro has update delay.

**Comment by helin24** [comment]
@xkeyC thanks for this report. I have a few clarifying questions:
1. I don't really understand how this http_proxy variable is supposed to work, as we don't refer to this within the plugin. Is this where you are expecting DevTools to run?
2. Is IntelliJ proceeding beyond the 'Starting DevTools' message before 30 seconds (30 sec is the automatic timeout)? If so, this would mean that DevTools is starting. But it sounds like this only works if you don't use http_proxy?
3. The error message you pasted about observatory - is this from the plugin after running your app has started? And is this when you don't use http_proxy?

**Comment by xkeyC** [comment]
> @xkeyC thanks for this report. I have a few clarifying questions:
> 
> 1. I don't really understand how this http_proxy variable is supposed to work, as we don't refer to this within the plugin. Is this where you are expecting DevTools to run?
> 2. Is IntelliJ proceeding beyond the 'Starting DevTools' message before 30 seconds (30 sec is the automatic timeout)? If so, this would mean that DevTools is starting. But it sounds like this only works if you don't use http_proxy?
> 3. The error message you pasted about observatory - is this from the plugin after running your app has started? And is this when you don't use http_proxy?

1. I hope flutter-intellij can read the proxy settings of idea/android-studio, and set http_proxy for dart when running flutter pub get, flutter pub upgrade, DevTools and other pub commands. (Just like gradle)
![image](https://user-images.githubusercontent.com/39891083/106643016-11b4f780-65c4-11eb-9914-a6b131408d6d.png)

2. If i use http_proxy, Dart DevTools will be initialized successfully (just like used outside mainland China),Then, the flutter app failed to run.

3. If I do not use http_proxy, the flutter app will start normally, but the dart DevTools will time out and will not work, and every time I press the run button, the "Starting DevTools" pops up.

In other words, if I use http_proxy, then I will not be able to start flutter app. If I don’t use http_proxy, then I will not be able to use dart DevTools.

**Comment by xkeyC** [comment]
I think there are two solutions:

1. flutter-intellij set the http_proxy environment variable at runtime for some operations.

2. dart pub adds a new http_proxy option, which can be a configuration file (like gradle) or a command line parameter (dart pub get --http_proxy="xxx" --https_proxy="xxxx"), if it is a configuration file , It will be edited by the user. If it is a command line parameter, it should be set by flutter-intellij.

**Comment by helin24** [comment]
Thanks for clarifying; I'm looking into how this works further. Are you able to set http_proxy and run your application through the command line? i.e. if you use `flutter run`, does this start the app and DevTools successfully?

**Comment by xkeyC** [comment]
> Thanks for clarifying; I'm looking into how this works further. Are you able to set http_proxy and run your application through the command line? i.e. if you use `flutter run`, does this start the app and DevTools successfully?

No,if use http proxy to `flutter run`,the dartvm will Fail to connect to service protocol.
```
[  +12 ms] E/DartVM  (30122): vm-service: Error: Unhandled exception:
[        ] E/DartVM  (30122): WebSocketException: Invalid WebSocket upgrade request
[        ] E/DartVM  (30122): 
[        ] E/flutter (30122): [ERROR:flutter/runtime/dart_isolate.cc(1137)] Unhandled exception:
[        ] E/flutter (30122): WebSocketException: Invalid WebSocket upgrade request
[ +109 ms] Fail to connect to service protocol: http://127.0.0.1:11498/Emoi_L7U-s8=/: HttpException: Connection closed before full header was received, uri = http://127.0.0.1:11498/Emoi_L7U-s8=/ws
```
I think that just setting the proxy for `flutter pub` command is the best solution. (start DevTools command is `lutter pub global activate devtools`,so itwill solve all problems.)

**Comment by UbuntuLover** [comment]
Hello I also met a wierd bug: 
I set up my dev environment as document described (my host is ubuntu 20.04 and using zsh). When I run `flutter doctor` it always prompts me `NO_PROXY does not contain 127.0.0.1`. However, I appended `export NO_PROXY=localhost,127.0.0.1` to my home directory's `.zshrc` file and sourced it. here is part of my `.zshrc` file:
```shell
export HTTP_PROXY=http://127.0.0.1:10877
export HTTPS_PROXY=http://127.0.0.1:10877
export NO_PROXY=localhost,127.0.0.1,LOCALHOST

```

while I run the hello world project I cant get the expected activity on emulator and same as using real device. The error is same as yours @xkeyC ..... Any tips for this? Thanks in advance!

你们好！ 我是一个刚刚入门flutter的小菜鸡，我按照官方文档进行搭建环境并且设置了相关环境变量，我的具体描述在上文里面。我遇到了一个跟楼主很类似的bug， 我跑文档的hello world 工程的时候白屏，然后提示报错就是（1137）的报错。。。。 有大哥有解决思路吗？ 谢谢各位啦~

**Comment by UbuntuLover** [comment]
> Hello I also met a wierd bug:
> I set up my dev environment as document described (my host is ubuntu 20.04 and using zsh). When I run `flutter doctor` it always prompts me `NO_PROXY does not contain 127.0.0.1`. However, I appended `export NO_PROXY=localhost,127.0.0.1` to my home directory's `.zshrc` file and sourced it. here is part of my `.zshrc` file:
> 
> ```shell
> export HTTP_PROXY=http://127.0.0.1:10877
> export HTTPS_PROXY=http://127.0.0.1:10877
> export NO_PROXY=localhost,127.0.0.1,LOCALHOST
> ```
> 
> while I run the hello world project I cant get the expected activity on emulator and same as using real device. The error is same as yours @xkeyC ..... Any tips for this? Thanks in advance!
> 
> 你们好！ 我是一个刚刚入门flutter的小菜鸡，我按照官方文档进行搭建环境并且设置了相关环境变量，我的具体描述在上文里面。我遇到了一个跟楼主很类似的bug， 我跑文档的hello world 工程的时候白屏，然后提示报错就是（1137）的报错。。。。 有大哥有解决思路吗？ 谢谢各位啦~

我应该继续补充的：Maybe I should add some details:
我运行`flutter doctor`这条命令的输出：Here is what I get by running `flutter doctor`
```shell
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✓] Chrome - develop for the web
[✓] Android Studio
[✓] IntelliJ IDEA Ultimate Edition (version 2020.2)
[✓] VS Code (version 1.51.1)
[!] Proxy Configuration
    ! NO_PROXY does not contain 127.0.0.1
[✓] Connected device (3 available)

! Doctor found issues in 1 category.
```

**Comment by xkeyC** [comment]
> > Hello I also met a wierd bug:
> > I set up my dev environment as document described (my host is ubuntu 20.04 and using zsh). When I run `flutter doctor` it always prompts me `NO_PROXY does not contain 127.0.0.1`. However, I appended `export NO_PROXY=localhost,127.0.0.1` to my home directory's `.zshrc` file and sourced it. here is part of my `.zshrc` file:
> > ```shell
> > export HTTP_PROXY=http://127.0.0.1:10877
> > export HTTPS_PROXY=http://127.0.0.1:10877
> > export NO_PROXY=localhost,127.0.0.1,LOCALHOST
> > ```
> > 
> > 
> > while I run the hello world project I cant get the expected activity on emulator and same as using real device. The error is same as yours @xkeyC ..... Any tips for this? Thanks in advance!
> > 你们好！ 我是一个刚刚入门flutter的小菜鸡，我按照官方文档进行搭建环境并且设置了相关环境变量，我的具体描述在上文里面。我遇到了一个跟楼主很类似的bug， 我跑文档的hello world 工程的时候白屏，然后提示报错就是（1137）的报错。。。。 有大哥有解决思路吗？ 谢谢各位啦~
> 
> 我应该继续补充的：Maybe I should add some details:
> 我运行`flutter doctor`这条命令的输出：Here is what I get by running `flutter doctor`
> 
> ```shell
> [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
> [✓] Chrome - develop for the web
> [✓] Android Studio
> [✓] IntelliJ IDEA Ultimate Edition (version 2020.2)
> [✓] VS Code (version 1.51.1)
> [!] Proxy Configuration
>     ! NO_PROXY does not contain 127.0.0.1
> [✓] Connected device (3 available)
> 
> ! Doctor found issues in 1 category.
> ```

Don't set http_proxy in your `.zshrc`,And export http_proxy when only using `flutter pub xxx`.

**Comment by UbuntuLover** [comment]
> > > Hello I also met a wierd bug:
> > > I set up my dev environment as document described (my host is ubuntu 20.04 and using zsh). When I run `flutter doctor` it always prompts me `NO_PROXY does not contain 127.0.0.1`. However, I appended `export NO_PROXY=localhost,127.0.0.1` to my home directory's `.zshrc` file and sourced it. here is part of my `.zshrc` file:
> > > ```shell
> > > export HTTP_PROXY=http://127.0.0.1:10877
> > > export HTTPS_PROXY=http://127.0.0.1:10877
> > > export NO_PROXY=localhost,127.0.0.1,LOCALHOST
> > > ```
> > > 
> > > 
> > > while I run the hello world project I cant get the expected activity on emulator and same as using real device. The error is same as yours @xkeyC ..... Any tips for this? Thanks in advance!
> > > 你们好！ 我是一个刚刚入门flutter的小菜鸡，我按照官方文档进行搭建环境并且设置了相关环境变量，我的具体描述在上文里面。我遇到了一个跟楼主很类似的bug， 我跑文档的hello world 工程的时候白屏，然后提示报错就是（1137）的报错。。。。 有大哥有解决思路吗？ 谢谢各位啦~
> > 
> > 
> > 我应该继续补充的：Maybe I should add some details:
> > 我运行`flutter doctor`这条命令的输出：Here is what I get by running `flutter doctor`
> > ```shell
> > [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
> > [✓] Chrome - develop for the web
> > [✓] Android Studio
> > [✓] IntelliJ IDEA Ultimate Edition (version 2020.2)
> > [✓] VS Code (version 1.51.1)
> > [!] Proxy Configuration
> >     ! NO_PROXY does not contain 127.0.0.1
> > [✓] Connected device (3 available)
> > 
> > ! Doctor found issues in 1 category.
> > ```
> 
> Don't set http_proxy in your `.zshrc`,And `export http_proxy` when only using `flutter pub xxx`.

OK， I removed HTTP_PROXY env variables. But `flutter doctor` always complain ` NO_PROXY does not contain 127.0.0.1` ..... I echo this variable in terminal and the value is correct.... QAQ

**Comment by UbuntuLover** [comment]
and this is the output of `flutter doctor` and `echo $NO_PROXY`...
```shell
➜  ~ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel master, 1.26.0-18.0.pre.212, on Linux, locale en_US.UTF-8)
⣽Error 1 retrieving device properties for ro.product.cpu.abi:
adb: insufficient permissions for device: user in plugdev group; are your udev
rules wrong?
See [http://developer.android.com/tools/device.html] for more information


[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
[✓] Chrome - develop for the web
[✓] Android Studio
[✓] IntelliJ IDEA Ultimate Edition (version 2020.2)
[✓] VS Code (version 1.51.1)
[!] Proxy Configuration
    ! NO_PROXY does not contain 127.0.0.1
[✓] Connected device (2 available)

! Doctor found issues in 1 category.
➜  ~ echo $NO_PROXY
localhost,127.0.0.1

```

What can I do right now ? Thanks for your reply~ @xkeyC

**Comment by xkeyC** [comment]
> and this is the output of `flutter doctor` and `echo $NO_PROXY`...
> 
> ```shell
> ➜  ~ flutter doctor
> Doctor summary (to see all details, run flutter doctor -v):
> [✓] Flutter (Channel master, 1.26.0-18.0.pre.212, on Linux, locale en_US.UTF-8)
> ⣽Error 1 retrieving device properties for ro.product.cpu.abi:
> adb: insufficient permissions for device: user in plugdev group; are your udev
> rules wrong?
> See [http://developer.android.com/tools/device.html] for more information
> 
> 
> [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
> [✓] Chrome - develop for the web
> [✓] Android Studio
> [✓] IntelliJ IDEA Ultimate Edition (version 2020.2)
> [✓] VS Code (version 1.51.1)
> [!] Proxy Configuration
>     ! NO_PROXY does not contain 127.0.0.1
> [✓] Connected device (2 available)
> 
> ! Doctor found issues in 1 category.
> ➜  ~ echo $NO_PROXY
> localhost,127.0.0.1
> ```
> 
> What can I do right now ? Thanks for your reply~ @xkeyC

I think you need remove 'NO_PROXY' too.'
Please set 'export http_proxy' when only using the ’flutter pub xxx’ command, if you use'flutter run', please do not.

**Comment by helin24** [comment]
@xkeyC the flutter tools team recommends setting `PUB_HOSTED_URL` instead of `HTTP_PROXY`: https://flutter.dev/community/china

Would this work for your case? The reason for this is `HTTP_PROXY` affects all http connections, but generally only the pub URL needs to be modified.

**Comment by xkeyC** [comment]
> @xkeyC the flutter tools team recommends setting `PUB_HOSTED_URL` instead of `HTTP_PROXY`: https://flutter.dev/community/china
> 
> Would this work for your case? The reason for this is `HTTP_PROXY` affects all http connections, but generally only the pub URL needs to be modified.

No,I'm using the master channel,the China mirro has update delay.

**Comment by helin24** [comment]
@xkeyC I see, sorry I don't have a more helpful suggestion here. I may be unclear on what the address `http://127.0.0.1:7890` in your example represents - if this is where pub is, then can you set `PUB_HOSTED_URL` equal to this instead of setting `HTTP_PROXY`?

We generally recommend using the dev channel (or beta, stable) and don't plan to make changes to support master for this case.

**Comment by xkeyC** [comment]
> @xkeyC I see, sorry I don't have a more helpful suggestion here. I may be unclear on what the address `http://127.0.0.1:7890` in your example represents - if this is where pub is, then can you set `PUB_HOSTED_URL` equal to this instead of setting `HTTP_PROXY`?
> 
> We generally recommend using the dev channel (or beta, stable) and don't plan to make changes to support master for this case.

The http://127.0.0.1:7890 is Http Proxy Url,this is a local conversion of other proxy protocols to support most applications.

**Comment by UbuntuLover** [comment]
> > and this is the output of `flutter doctor` and `echo $NO_PROXY`...
> > ```shell
> > ➜  ~ flutter doctor
> > Doctor summary (to see all details, run flutter doctor -v):
> > [✓] Flutter (Channel master, 1.26.0-18.0.pre.212, on Linux, locale en_US.UTF-8)
> > ⣽Error 1 retrieving device properties for ro.product.cpu.abi:
> > adb: insufficient permissions for device: user in plugdev group; are your udev
> > rules wrong?
> > See [http://developer.android.com/tools/device.html] for more information
> > 
> > 
> > [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
> > [✓] Chrome - develop for the web
> > [✓] Android Studio
> > [✓] IntelliJ IDEA Ultimate Edition (version 2020.2)
> > [✓] VS Code (version 1.51.1)
> > [!] Proxy Configuration
> >     ! NO_PROXY does not contain 127.0.0.1
> > [✓] Connected device (2 available)
> > 
> > ! Doctor found issues in 1 category.
> > ➜  ~ echo $NO_PROXY
> > localhost,127.0.0.1
> > ```
> > 
> > 
> > What can I do right now ? Thanks for your reply~ @xkeyC
> 
> I think you need remove `NO_PROXY` too.
> Please set 'export http_proxy' when only using the `flutter pub xxx` command, if you use `flutter run`, please do not.

Thanks for reply and everything works... THANK YOU VERY MUCH!!!

**Comment by helin24** [comment]
@xkeyC We're not planning to build support for this in IntelliJ, but I'm hoping we can figure out a workaround for you and gather some information so we understood your use case better. 

The first issue is we're not familiar with how the HTTP proxy URL works, but I'm gathering this is similar to what's described [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Proxy_servers_and_tunneling)? Essentially, the URL `http://127.0.0.1:7890` is not the location of pub itself, but this is the proxy through which you access the normal pub.dev site and any other address. Is that correct?

Regarding `PUB_HOSTED_URL`, I understand that you can't set this to a URL that is fully up to date with master, but would it work for your purposes to stay on flutter master, but set `PUB_HOSTED_URL` to one of the trailing mirror sites? This way DevTools could be started (potentially with a slightly earlier version, but may be the same), but you won't have the problem of the VM service URL being inaccessible through a proxy URL.

I'm also curious why you're using flutter master instead of one of the better supported channels. Do you have a use case where dev/beta/stable flutter does not suffice?

**Comment by xkeyC** [comment]
@helin24 
> the URL http://127.0.0.1:7890 is not the location of pub itself, but this is the proxy through which you access the normal pub.dev site and any other address. Is that correct?
yes.
>I'm also curious why you're using flutter master instead of one of the better supported channels. Do you have a use case where dev/beta/stable flutter does not suffice?
Just for fun, and discover bugs. (https://github.com/flutter/flutter/issues/64864)

**Comment by xkeyC** [comment]
> Regarding PUB_HOSTED_URL, I understand that you can't set this to a URL that is fully up to date with master, but would it work for your purposes to stay on flutter master, but set PUB_HOSTED_URL to one of the trailing mirror sites? This way DevTools could be started (potentially with a slightly earlier version, but may be the same), but you won't have the problem of the VM service URL being inaccessible through a proxy URL.

Sometimes, Chinese pub mirro can't compile the master channel's flutter.

**Comment by xkeyC** [comment]
So, Http_Proxy is the only way I can use the Master channel.

If i use a relatively new package, the Maven mirro in China may have the problem of update delay. I think Http_Proxy is the best solution.

Gradle, maven....they all have good HTTP proxy support. Why don't dart pub do it?

**Comment by helin24** [comment]
@xkeyC interesting, so the pub mirror site does not only provide the DevTools package, it also compiles the local version of flutter you are running? 

I'm going to close this issue since we're not going to be making changes to the plugin to accommodate this case (although we are looking into app startup changes so it will not hang on DevTools starting). I recommend potentially posting to flutter/flutter with suggestions on how HTTP_PROXY or PUB_HOSTED_URL could be changed for this case; someone there may also have better recommendations for workarounds.

**Comment by xkeyC** [comment]
> @xkeyC interesting, so the pub mirror site does not only provide the DevTools package, it also compiles the local version of flutter you are running?
> 
> I'm going to close this issue since we're not going to be making changes to the plugin to accommodate this case (although we are looking into app startup changes so it will not hang on DevTools starting). I recommend potentially posting to flutter/flutter with suggestions on how HTTP_PROXY or PUB_HOSTED_URL could be changed for this case; someone there may also have better recommendations for workarounds.

![image](https://user-images.githubusercontent.com/39891083/107487618-5262d100-6bc1-11eb-8aeb-9b74778bfca0.png)
![image](https://user-images.githubusercontent.com/39891083/107487644-5b53a280-6bc1-11eb-85f5-4f803252edde.png)


Okay,I think I can use my own method to solve this problem: https://github.com/clinux-co/proxy2pub .
It can convert general `Http_Proxy` into `FLUTTER_STORAGE_BASE_URL` and `PUB_HOSTED_URL.`
@helin24  @UbuntuLover

**Comment by UbuntuLover** [comment]
yeah. Thanks bro.
